llvm/lib/Target/RISCV/RISCVInstrInfo.cpp

```diff
@@ -2735,6 +2779,100 @@ RISCVInstrInfo::getSerializableMachineMemOperandTargetFlags() const {
   return ArrayRef(TargetFlags);
 }
 
+std::pair<Register, Register>
+RISCVInstrInfo::adjustVLVTYPE(MachineInstr &MI,
+                              MachineBasicBlock &MBB,
+                              unsigned SEW, unsigned LMUL) const {
+  // Save origin VL and VTYPE, then set VL to VLMAX, and adjust
+  // SEW and LMUL according to the parameters in front of the MI.
+  auto DL = MI.getDebugLoc();
+  auto *MRI = &MBB.getParent()->getRegInfo();
+
+  Register SavedVL = MRI->createVirtualRegister(&RISCV::GPRRegClass);
+  Register SavedVType = MRI->createVirtualRegister(&RISCV::GPRRegClass);
+
+  // Spec: The assembler pseudoinstruction to read a CSR, `CSRR rd, csr`, is
+  // encoded as `CSRRS rd, csr, x0`.
+  BuildMI(MBB, MI, DL, get(RISCV::CSRRS), SavedVL)
+      .addImm(RISCVSysReg::lookupSysRegByName("VL")->Encoding)
+      .addReg(RISCV::X0);
+  BuildMI(MBB, MI, DL, get(RISCV::CSRRS), SavedVType)
+      .addImm(RISCVSysReg::lookupSysRegByName("VTYPE")->Encoding)
+      .addReg(RISCV::X0);
+
+  auto VTypeI = RISCVVType::encodeXTHeadVTYPE(SEW, LMUL, 1);
+  BuildMI(MBB, MI, DL, get(RISCV::TH_VSETVLI))
```

Comment:
I think `PseudoTH_VSETVLIX0` is theoretically better than `TH_VSETVLI`. Some internal passes work only with the pseudo version like `isConvertibleToVMV`. But that's up to you 😉

Reply:
I suddenly realized that maybe it is inappropriate to use `TH_VSETVLI` here. I did the modification because the `RISCVInsertVSETVLI` pass will check if a `PseudoVSETVLI` instruction is of the form `vsetvli x0, x0, vtype` and abort if it is. So if we use the Pseudo version some of our tests will crash.

But in RVV 0.7.1 the meaning of `vsetvli x0, x0, vtype` has subtle difference with that in RVV 1.0. I will check if we can use the pseudo version for better performance here.

Thank you for pointing it out!